src/java/org/apache/cassandra/service/reads/trackwarnings/CoordinatorTrackWarnings.java

```diff
@@ -0,0 +1,182 @@
+        if (STATE.get() != null)
+        {
+            if (ENABLE_DEFENSIVE_CHECKS)
+                throw new AssertionError("CoordinatorTrackWarnings.init called while state is not null: " + STATE.get());
+            return;
+        }
+        STATE.set(INIT);
+    }
+
+    public static void update(ReadCommand cmd, TrackWarningsSnapshot snapshot)
+    {
+        logger.trace("CoordinatorTrackWarnings.update({}, {})", cmd.metadata(), snapshot);
+        Map<ReadCommand, TrackWarningsSnapshot> map = mutable();
+        TrackWarningsSnapshot previous = map.get(cmd);
+        TrackWarningsSnapshot update = TrackWarningsSnapshot.merge(previous, snapshot);
+        if (update == null) // null happens when the merge had null input or EMPTY input... remove the command from the map
+            map.remove(cmd);
+        else
+            map.put(cmd, update);
+    }
+
+    public static void done()
+    {
+        Map<ReadCommand, TrackWarningsSnapshot> map = readonly();
+        logger.trace("CoordinatorTrackWarnings.done() with state {}", map);
+        map.forEach((command, merged) -> {
+            ColumnFamilyStore cfs = Schema.instance.getColumnFamilyStoreInstance(command.metadata().id);
+            // race condition when dropping tables, also happens in unit tests as Schema may be bypassed
+            if (cfs == null)
+                return;
+
+            String cql = command.toCQLString();
+            String loggableTokens = command.loggableTokens();
+            trackAborts(merged.tombstones, cql, loggableTokens, cfs.metric.clientTombstoneAborts, ReadCallback::tombstoneAbortMessage);
+            trackWarnings(merged.tombstones, cql, loggableTokens, cfs.metric.clientTombstoneWarnings, ReadCallback::tombstoneWarnMessage);
+
+            trackAborts(merged.localReadSize, cql, loggableTokens, cfs.metric.localReadSizeAborts, ReadCallback::localReadSizeAbortMessage);
+            trackWarnings(merged.localReadSize, cql, loggableTokens, cfs.metric.localReadSizeWarnings, ReadCallback::localReadSizeWarnMessage);
+
+            trackAborts(merged.rowIndexTooLarge, cql, loggableTokens, cfs.metric.rowIndexSizeAborts, ReadCallback::rowIndexSizeAbortMessage);
+            trackWarnings(merged.rowIndexTooLarge, cql, loggableTokens, cfs.metric.rowIndexSizeWarnings, ReadCallback::rowIndexSizeWarnMessage);
+        });
+
+        STATE.set(null);
+    }
+
+    private static Map<ReadCommand, TrackWarningsSnapshot> mutable()
+    {
+        Map<ReadCommand, TrackWarningsSnapshot> map = STATE.get();
+        if (map == null)
+        {
+            if (ENABLE_DEFENSIVE_CHECKS)
+                throw new AssertionError("CoordinatorTrackWarnings.mutable calling without calling .init() first");
+            // set map to an "ignore" map; dropping all mutations
+            // this does not update the state as it is not known if .clear() will be called; so safter to leave state null
+            map = IgnoreMap.get();
```

Comment:
Would it be simpler to just always make the assertion (like a literal `assert`) here and get rid of the `IgnoreMap` concept? I guess when I was worried about this earlier, it was the overhead of the perhaps unnecessary map creations that I was worried about, not the `asserts`.

Reply:
Same comment for `readonly()` below.

Reply:
This also resolves some oddities like `mutable()` returning an immutable map in some cases ;)

Reply:
the argument for Ignore is to not fail code paths which call outside of the thread.  So if you enable failures for this feature we block, else we silently ignore.  The issue with `assert` is it is NEVER safe to disable in C*... so you can never opt out of failing in this case.

My thinking is really just that, if we have a bug in C* why fail in production?  The user negative to silently ignore is we do not update logs/metrics, but their queries still work; why fail the query as well?

Reply:
tl;dr being enabled in all tests is good enough for me